// @oneblink/sdk/tenants/classes/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendEmail = exports.DataManager = exports.EmailTemplates = exports.FormsAppEnvironments = exports.Approvals = exports.PDF = exports.TeamMembers = exports.Organisations = exports.Keys = exports.Jobs = exports.FormsApps = exports.Forms = void 0;
// Classes
const Forms_1 = __importDefault(require("./Forms"));
exports.Forms = Forms_1.default;
const FormsApps_1 = __importDefault(require("./FormsApps"));
exports.FormsApps = FormsApps_1.default;
const Jobs_1 = __importDefault(require("./Jobs"));
exports.Jobs = Jobs_1.default;
const Keys_1 = __importDefault(require("./Keys"));
exports.Keys = Keys_1.default;
const Organisations_1 = __importDefault(require("./Organisations"));
exports.Organisations = Organisations_1.default;
const TeamMembers_1 = __importDefault(require("./TeamMembers"));
exports.TeamMembers = TeamMembers_1.default;
const PDF_1 = __importDefault(require("./PDF"));
exports.PDF = PDF_1.default;
const Approvals_1 = __importDefault(require("./Approvals"));
exports.Approvals = Approvals_1.default;
const FormsAppEnvironments_1 = __importDefault(require("./FormsAppEnvironments"));
exports.FormsAppEnvironments = FormsAppEnvironments_1.default;
const EmailTemplates_1 = __importDefault(require("./EmailTemplates"));
exports.EmailTemplates = EmailTemplates_1.default;
const DataManager_1 = __importDefault(require("./DataManager"));
exports.DataManager = DataManager_1.default;
// Functions
const sendEmail_1 = __importDefault(require("./sendEmail"));
exports.sendEmail = sendEmail_1.default;
