// @oneblink/sdk-core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fileUploadService = exports.schedulingService = exports.paymentService = exports.abnService = exports.userService = exports.typeCastService = exports.submissionService = exports.conditionalLogicService = exports.formElementsService = void 0;
/**
 * ## Form Elements Service
 *
 * Helper functions for handling form elements
 *
 * ```js
 * import { formElementsService } from '@oneblink/sdk-core'
 * ```
 */
exports.formElementsService = __importStar(require("./formElementsService"));
/**
 * ## Conditional Logic Service
 *
 * Helper functions for handling conditional logic
 *
 * ```js
 * import { conditionalLogicService } from '@oneblink/sdk-core'
 * ```
 */
exports.conditionalLogicService = __importStar(require("./conditionalLogicService"));
/**
 * ## Submission Service
 *
 * Helper functions for handling form submission data
 *
 * ```js
 * import { submissionService } from '@oneblink/sdk-core'
 * ```
 */
exports.submissionService = __importStar(require("./submissionService"));
/**
 * ## Submission Service
 *
 * Helper functions for casting types
 *
 * ```js
 * import { typeCastService } from '@oneblink/sdk-core'
 * ```
 */
exports.typeCastService = __importStar(require("./typeCastService"));
/**
 * ## User Service
 *
 * Helper functions for handling users
 *
 * ```js
 * import { userService } from '@oneblink/sdk-core'
 * ```
 */
exports.userService = __importStar(require("./userService"));
/**
 * ## ABN Service
 *
 * Helper functions for handling ABN Records
 *
 * ```js
 * import { abnService } from '@oneblink/sdk-core'
 * ```
 */
exports.abnService = __importStar(require("./abnService"));
/**
 * ## Payment Service
 *
 * Helper functions for payments
 *
 * ```js
 * import { paymentService } from '@oneblink/sdk-core'
 * ```
 */
exports.paymentService = __importStar(require("./paymentService"));
/**
 * ## Scheduling Service
 *
 * Helper functions for scheduling
 *
 * ```js
 * import { schedulingService } from '@oneblink/sdk-core'
 * ```
 */
exports.schedulingService = __importStar(require("./schedulingService"));
/**
 * ## File Upload Service
 *
 * Helper functions for handling file uploads
 *
 * ```js
 * import { fileUploadService } from '@oneblink/sdk-core'
 * ```
 */
exports.fileUploadService = __importStar(require("./fileUploadService"));
