// @oneblink/sdk-core/dist/schedulingService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkForSchedulingEvent = void 0;
const _1 = require(".");
/**
 * Examine a submission and its form definition to validate whether a scheduling
 * workflow event needs to run.
 *
 * #### Example
 *
 * ```js
 * const result = schedulingService.checkFormSchedulingEvent(
 *   form,
 *   submission,
 * )
 * ```
 *
 * @param definition
 * @param submission
 * @returns
 */
function checkForSchedulingEvent(definition, submission) {
    const schedulingSubmissionEvents = definition.schedulingEvents || [];
    return schedulingSubmissionEvents.find((schedulingSubmissionEvent) => _1.conditionalLogicService.evaluateConditionalPredicates({
        isConditional: !!schedulingSubmissionEvent.conditionallyExecute,
        requiresAllConditionalPredicates: !!schedulingSubmissionEvent.requiresAllConditionallyExecutePredicates,
        conditionalPredicates: schedulingSubmissionEvent.conditionallyExecutePredicates || [],
        submission: submission,
        formElements: definition.elements,
    }));
}
exports.checkForSchedulingEvent = checkForSchedulingEvent;
