// @oneblink/sdk/tenants/lib/pre-fill-data.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const aws_sdk_1 = __importDefault(require("aws-sdk"));
async function setPreFillData(preFillMeta, data) {
    const s3 = new aws_sdk_1.default.S3({
        region: preFillMeta.s3.region,
        accessKeyId: preFillMeta.credentials.AccessKeyId,
        secretAccessKey: preFillMeta.credentials.SecretAccessKey,
        sessionToken: preFillMeta.credentials.SessionToken,
    });
    const objectMeta = {
        Bucket: preFillMeta.s3.bucket,
        Key: preFillMeta.s3.key,
        Body: JSON.stringify(data),
        Expires: new Date(new Date().setFullYear(new Date().getFullYear() + 1)),
        CacheControl: 'max-age=31536000',
        ContentType: 'application/json',
    };
    const uploadOptions = {
        partSize: 5 * 1024 * 1024,
        queueSize: 1,
    };
    await s3.upload(objectMeta, uploadOptions).promise();
}
exports.default = setPreFillData;
