// @oneblink/sdk/tenants/lib/upload-asset.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const aws_sdk_1 = __importDefault(require("aws-sdk"));
async function uploadAsset(assetCredentials, data, assetContentType) {
    const s3 = new aws_sdk_1.default.S3({
        region: assetCredentials.s3.region,
        accessKeyId: assetCredentials.credentials.AccessKeyId,
        secretAccessKey: assetCredentials.credentials.SecretAccessKey,
        sessionToken: assetCredentials.credentials.SessionToken,
    });
    const objectMeta = {
        Bucket: assetCredentials.s3.bucket,
        Key: assetCredentials.s3.key,
        Body: data,
        Expires: new Date(new Date().setFullYear(new Date().getFullYear() + 1)),
        CacheControl: 'max-age=31536000',
        ContentType: assetContentType,
        ACL: 'public-read',
    };
    const uploadOptions = {
        partSize: 5 * 1024 * 1024,
        queueSize: 1,
    };
    return await s3.upload(objectMeta, uploadOptions).promise();
}
exports.default = uploadAsset;
