// @oneblink/sdk/tenants/lib/forms-schema/elements/CalculationElement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.type = void 0;
const joi_1 = __importDefault(require("joi"));
const property_schemas_1 = require("../property-schemas");
const common_1 = require("../common");
exports.type = 'calculation';
exports.default = joi_1.default.object({
    ...property_schemas_1.baseSchemas,
    name: property_schemas_1.name,
    label: property_schemas_1.label,
    ...property_schemas_1.requiredSchemas,
    readOnly: property_schemas_1.readOnly,
    ...property_schemas_1.conditionallyShowSchemas,
    defaultValue: joi_1.default.string().required().regex(common_1.base64DataRegex, {
        name: 'No Binary Data',
        invert: true,
    }),
    calculation: joi_1.default.string().required(),
    preCalculationDisplay: joi_1.default.string()
        .allow(null)
        .regex(common_1.base64DataRegex, {
        name: 'No Binary Data',
        invert: true,
    }),
    displayAsCurrency: joi_1.default.boolean().allow(null),
    customCssClasses: property_schemas_1.customCssClasses,
});
