// @oneblink/sdk/tenants/lib/forms-schema/elements/FreshdeskDependentFieldElement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.type = void 0;
const joi_1 = __importDefault(require("joi"));
const property_schemas_1 = require("../property-schemas");
exports.type = 'freshdeskDependentField';
exports.default = joi_1.default.object({
    ...property_schemas_1.baseSchemas,
    name: property_schemas_1.name,
    label: property_schemas_1.label,
    hint: property_schemas_1.hint,
    subCategoryLabel: property_schemas_1.label,
    subCategoryHint: property_schemas_1.hint,
    itemLabel: property_schemas_1.label,
    itemHint: property_schemas_1.hint,
    ...property_schemas_1.requiredSchemas,
    readOnly: property_schemas_1.readOnly,
    ...property_schemas_1.conditionallyShowSchemas,
    ...property_schemas_1.lookupSchemas,
    defaultValue: joi_1.default.object().keys({
        category: joi_1.default.string(),
        subCategory: joi_1.default.string(),
        item: joi_1.default.string(),
    }),
    ...property_schemas_1.optionsSchemas,
    customCssClasses: property_schemas_1.customCssClasses,
});
