// @oneblink/sdk/tenants/lib/forms-schema/elements/PointElement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.type = void 0;
const joi_1 = __importDefault(require("joi"));
const property_schemas_1 = require("../property-schemas");
exports.type = 'pointAddress';
exports.default = joi_1.default.object({
    ...property_schemas_1.baseSchemas,
    name: property_schemas_1.name,
    label: property_schemas_1.label,
    hint: property_schemas_1.hint,
    ...property_schemas_1.requiredSchemas,
    readOnly: property_schemas_1.readOnly,
    ...property_schemas_1.conditionallyShowSchemas,
    ...property_schemas_1.lookupSchemas,
    placeholderValue: property_schemas_1.placeholderValue,
    stateTerritoryFilter: joi_1.default.array().items(joi_1.default.string()),
    addressTypeFilter: joi_1.default.array().items(joi_1.default.string()),
    customCssClasses: property_schemas_1.customCssClasses,
});
