// @oneblink/sdk/tenants/lib/forms-schema/elements/RepeatableSetElement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.type = void 0;
const joi_1 = __importDefault(require("joi"));
const property_schemas_1 = require("../property-schemas");
const element_schema_1 = __importDefault(require("../element-schema"));
// Think this needs to be a variable because of recursive dependency
exports.type = 'repeatableSet';
const schema = joi_1.default.object({
    ...property_schemas_1.baseSchemas,
    name: property_schemas_1.name,
    label: property_schemas_1.label,
    hint: property_schemas_1.hint,
    readOnly: property_schemas_1.readOnly,
    ...property_schemas_1.conditionallyShowSchemas,
    minSetEntries: joi_1.default.number().min(0),
    maxSetEntries: joi_1.default.number().when('minSetEntries', {
        is: joi_1.default.number().required().min(0),
        then: joi_1.default.number().min(joi_1.default.ref('minSetEntries', { render: true })),
        otherwise: joi_1.default.number().min(0),
    }),
    addSetEntryLabel: joi_1.default.string(),
    removeSetEntryLabel: joi_1.default.string(),
    elements: joi_1.default.array()
        .items(joi_1.default.custom((value) => {
        if (!value)
            return;
        const result = element_schema_1.default.validate(value);
        if (result.error) {
            throw result.error;
        }
        return result.value;
    }))
        .required()
        .min(1)
        .unique('name', { ignoreUndefined: true })
        .unique('id'),
    customCssClasses: property_schemas_1.customCssClasses,
});
exports.default = schema;
