// @oneblink/sdk/tenants/lib/forms-schema/elements/SelectElement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.type = void 0;
const joi_1 = __importDefault(require("joi"));
const property_schemas_1 = require("../property-schemas");
exports.type = 'select';
exports.default = joi_1.default.object({
    ...property_schemas_1.baseSchemas,
    name: property_schemas_1.name,
    label: property_schemas_1.label,
    hint: property_schemas_1.hint,
    ...property_schemas_1.requiredSchemas,
    readOnly: property_schemas_1.readOnly,
    ...property_schemas_1.conditionallyShowSchemas,
    ...property_schemas_1.lookupSchemas,
    multi: joi_1.default.boolean(),
    defaultValue: joi_1.default.when('multi', {
        is: true,
        then: property_schemas_1.defaultValueOptionsMultiple,
        otherwise: property_schemas_1.defaultValueOptionsSingle,
    }),
    buttons: property_schemas_1.buttons,
    ...property_schemas_1.optionsSchemas,
    canToggleAll: joi_1.default.when('multi', {
        is: true,
        then: property_schemas_1.canToggleAll,
        otherwise: joi_1.default.any().strip(),
    }),
    customCssClasses: property_schemas_1.customCssClasses,
});
