// @oneblink/sdk/tenants/lib/forms-schema/elements/TextElement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.textareaElementType = exports.textElementType = void 0;
const joi_1 = __importDefault(require("joi"));
const property_schemas_1 = require("../property-schemas");
exports.textElementType = 'text';
exports.textareaElementType = 'textarea';
exports.default = joi_1.default.object({
    ...property_schemas_1.baseSchemas,
    name: property_schemas_1.name,
    label: property_schemas_1.label,
    hint: property_schemas_1.hint,
    ...property_schemas_1.requiredSchemas,
    readOnly: property_schemas_1.readOnly,
    ...property_schemas_1.conditionallyShowSchemas,
    ...property_schemas_1.lookupSchemas,
    placeholderValue: property_schemas_1.placeholderValue,
    minLength: joi_1.default.number().min(0),
    maxLength: joi_1.default.number()
        .min(0)
        .when('minLength', {
        is: joi_1.default.number().required(),
        then: joi_1.default.number().min(joi_1.default.ref('minLength', { render: true })),
    }),
    defaultValue: joi_1.default.when('minLength', {
        is: joi_1.default.number().required(),
        then: joi_1.default.string().when('maxLength', {
            is: joi_1.default.number().required(),
            then: joi_1.default.string()
                .min(joi_1.default.ref('minLength', { render: true }))
                .max(joi_1.default.ref('maxLength', { render: true })),
            otherwise: joi_1.default.string().min(joi_1.default.ref('minLength', { render: true })),
        }),
    }).when('maxLength', {
        is: joi_1.default.number().required(),
        then: joi_1.default.string().max(joi_1.default.ref('maxLength', { render: true })),
        otherwise: joi_1.default.string(),
    }),
    ...property_schemas_1.regexSchemas,
    customCssClasses: property_schemas_1.customCssClasses,
});
